google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -459,22 +452,23 @@ StatusOr<std::unique_ptr<SessionHolder>> ConnectionImpl::GetSession(
     if (!response) {
       return response.status();
     }
-    session_name = std::move(*response->mutable_name());
+    session = google::cloud::internal::make_unique<Session>(
+        std::move(*response->mutable_name()));
   }
 
+  if (dissociate_from_pool) {
+    // Uses the default deleter; the Session is not returned to the pool.
+    return {std::move(session)};
+  }
   // TODO(#409) take a (weak) reference to `this` to avoid use-after-free.
-  auto deleter = dissociate_from_pool
-                     ? std::function<void(std::string)>()
-                     : [this](std::string session) {
-                         this->ReleaseSession(std::move(session));
-                       };
-  return ::google::cloud::internal::make_unique<SessionHolder>(
-      std::move(session_name), std::move(deleter));
+  return SessionHolder(session.release(), [this](Session* session) {
+    this->ReleaseSession(session);
+  });
 }
 
-void ConnectionImpl::ReleaseSession(std::string session) {
+void ConnectionImpl::ReleaseSession(Session* session) {
   std::lock_guard<std::mutex> lk(mu_);
-  sessions_.push_back(std::move(session));
+  sessions_.emplace_back(session);
```

Comment:
Maybe still `push_back`? https://abseil.io/tips/112

Reply:
I think `std::unique_ptr<T>::unique_ptr(T*)` is `explicit`, so `push_back()` does not work in this case?

Reply:
I missed that sessions_ was a `vector<unique_ptr<T>>`. So you and Todd are right here. My mistake.